src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
  * actually transmitted matched a pre-computed size and hash
  */
 class Runner {
-  constructor(agent, agentOpts) {
-    this.agent = agent || new https.Agent(agentOpts || {}); 
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      agent: Joi.object(),
+      agentOpts: Joi.object(),
+      maxRetries: Joi.number().min(0).max(10).default(5),
+    }).without('agent', 'agentOpts').optionalKeys(['agent', 'agentOpts']));
+
+    this.agent = opts.agent || new https.Agent(opts.agentOpts || {}); 
+
+    this.maxRetries = opts.maxRetries;
+
   }
 
+  /**
+   * Run a request and retry if a retryable error occurs.  Only 500-series
+   * responses are considered retriable, so any other status code will return
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
```

Comment:
I think we should drop stream support...The stream factory you were talking about we should keep, but with a stream factory we can do retries...

Note: the node FD is the same as a stream factory... I think node manages seek position inside the stream.... Look at the FS api, there is no seek method.. it done with createReadStream which can take a path or an fd